### PR TITLE
Fix proxy sanitization breaking class methods with `#private` fields

### PR DESCRIPTION
**Pull request by dahlia**
## Summary

Fixes the deferred-prompt-value sanitization proxy so that class-based parser results with private fields can be used safely inside config loaders (`load()`, `getConfigPath()`).

- Non-plain objects containing deferred prompt values are now sanitized by temporarily replacing deferred own properties on the target instance, calling the method, and restoring the originals afterward. This allows `this.#field` to work (since `this` is the real instance) while `this.publicField` reads the sanitized value.
- Async methods (`async` functions and Promise-returning methods) keep properties sanitized until the returned Promise settles, with reference counting to handle overlapping concurrent calls.
- The proxy's `get` trap still intercepts property reads for non-method access, `getOwnPropertyDescriptor` returns sanitized descriptors, and method wrappers are cached per key with invalidation when the underlying function changes.
- Inherited accessor (getter) descriptors are detected by walking the prototype chain, so getter-returned functions are not incorrectly cached.
- When a method is extracted and rebound via `call`/`apply`/`bind` (`this !== proxy`), the wrapper respects the caller-supplied receiver.

## Known limitations

The temporary-mutation approach has inherent trade-offs due to JavaScript's fundamental limitation that private fields cannot be accessed through a Proxy:

- **Observable mutation window:** During a method call, other aliases of the same instance see the temporarily sanitized property values. For async methods this window extends until the Promise settles.
- **`this` identity:** Methods run with the original target as `this`, not the proxy. Code that compares `this` against the sanitized object (e.g., as a `WeakMap` key) will observe a different identity.
- **Explicit writes matching sanitized values:** If a method intentionally writes the same value that the sanitizer injected (e.g., `this.token = undefined` on a field that was sanitized to `undefined`), the restoration logic cannot distinguish this from "untouched" and restores the original. This is a narrow edge case since the sanitized value is typically `undefined` and the method would be writing `undefined` to a field that already appeared to be `undefined`.
- **Frozen/sealed objects:** If own properties cannot be temporarily replaced (`Object.defineProperty` throws), the sanitized-target path returns `SANITIZE_FAILED` and the method falls back to the proxy receiver, where private-field access will propagate a `TypeError`.

These limitations are documented here because they are intrinsic to the only viable approach for supporting private fields in a sanitizing proxy. Addressing them would require language-level changes to how private fields interact with Proxies.

Closes https://github.com/dahlia/optique/issues/307

## Test plan

- [x] `mise check` passes (type check, lint, format, dry-run publish)
- [x] `mise test` passes across all runtimes (Deno, Node.js, Bun) with 354 tests
- [ ] Verify the following regression tests cover the reported scenario:
  - Private field access through sanitized proxy (`getSecret()`)
  - Deferred value stripping on public fields (`proxy.deferred === undefined`)
  - Mixed private + public field methods (`hasToken()` with `this.#id` and `this.token`)
  - Async methods with private fields (before and after `await`)
  - Non-retried application `TypeError`s (including messages containing "private")
  - Frozen object graceful fallback
  - Overlapping concurrent async calls with reference counting
  - Rebound method receiver preservation (`greetFn.call(other)`)
  - Method wrapper identity stability (`parsed.method === parsed.method`)
  - Promise-returning non-async methods with private fields

### CODE REVIEW

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 1]
```diff
@@ -317,20 +317,249 @@ function finalizeParsedForContext(
...
+    strip,
+    seen,
+  );
+  if (result !== SANITIZE_FAILED) return result;
+
+  // SANITIZE_FAILED means the target is frozen/sealed and its properties
+  // cannot be temporarily replaced.  Fall back to the proxy path, which
+  // handles methods that don't access private fields.  Private-field
+  // methods on frozen objects will propagate the TypeError.
+  return strip(fn.apply(proxy, args), seen);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Fixed in d19842f. The SANITIZE_FAILED fallback now checks for `instanceof Promise` and attaches `.then()` to strip the resolved value.

File: packages/config/src/index.ts [hunk 2]
```diff
@@ -183,20 +183,225 @@ function containsDeferredPromptValues(
...
+    fn,
+    args,
+    strip,
+    seen,
+  );
+  if (result !== SANITIZE_FAILED) return result;
+
+  // SANITIZE_FAILED means the target is frozen/sealed and its properties
+  // cannot be temporarily replaced.  Fall back to the proxy path.
+  return strip(fn.apply(proxy, args), seen);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts]: Fixed in d19842f. Same fix applied to the config path.

File: packages/config/src/run.test.ts [hunk 3]
```diff
@@ -1253,4 +1253,630 @@ describe("run with config context", { concurrency: false }, () => {
...
+    const injectContext: SourceContext = {
+      id: Symbol.for("@test/inject-class-instance"),
+      mode: "dynamic",
+      getAnnotations(parsed?: unknown) {
+        if (parsed == null) return {};
+        // Replace the parsed value with a class instance containing a
+        // deferred prompt value.  This simulates what happens when map()
+        // transforms a prompt-backed config field into a class instance.
+        return {};
+      },
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/run.test.ts]: Fixed in d19842f. Removed the unused `injectContext` and `runWith` setup, keeping only the direct `context.getAnnotations()` call.

File: packages/core/src/facade.ts [hunk 4]
```diff
@@ -317,20 +317,249 @@ function finalizeParsedForContext(
...
 ): T {
-  // FIXME: Proxying non-plain parsed values changes method receiver semantics,
-  // and method-only wrappers still need a safe way to expose scrubbed values.
+  // NOTE: Methods are invoked with the proxy as receiver so that property
+  // reads inside the method body go through the sanitized get trap.  If the
+  // method accesses private fields, that call throws a TypeError because
+  // private fields are bound to the original instance, not the proxy; in
+  // that case we temporarily sanitize the target's own properties, retry
+  // with the target as receiver, and restore the original values afterward.
   // See: https://github.com/dahlia/optique/issues/407
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Fixed in d19842f. Updated to reference #307.

File: packages/config/src/index.ts [hunk 5]
```diff
@@ -183,20 +183,225 @@ function containsDeferredPromptValues(
...
 ): T {
-  // FIXME: This proxy changes method receiver semantics and still cannot scrub
-  // deferred prompt values hidden behind private fields or wrapper methods.
+  // NOTE: Methods are invoked with the proxy as receiver so that property
+  // reads inside the method body go through the sanitized get trap.  If the
+  // method accesses private fields, that call throws a TypeError because
+  // private fields are bound to the original instance, not the proxy; in
+  // that case we temporarily sanitize the target's own properties, retry
+  // with the target as receiver, and restore the original values afterward.
   // See: https://github.com/dahlia/optique/issues/407
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts]: Fixed in d19842f. Updated to reference #307.

File: CHANGES.md [hunk 6]
```diff
@@ -297,6 +297,11 @@ To be released.
     `bindConfig()` from resolving config values through wrapper combinators.
     [[#385], [#535]]
 
+ -  Fixed proxy-based sanitization of deferred prompt values breaking class
+    methods that access private fields.  Methods on non-plain objects are now
+    bound to the original instance so private field access works correctly
+    through the sanitized view.  [[#307], [#558]]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md]: Fixed in d19842f. Reworded to describe the temporary own-property sanitization strategy.

File: CHANGES.md [hunk 7]
```diff
@@ -422,6 +430,11 @@ To be released.
     `undefined`, while keeping the original parsed object identity when no
     sanitization is needed.  [[#177], [#490]]
 
+ -  Fixed proxy-based sanitization of deferred prompt values breaking class
+    methods that access private fields.  Methods on non-plain objects are now
+    bound to the original instance so private field access works correctly
+    through the sanitized view.  [[#307], [#558]]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md]: Fixed in d19842f. Same rewording applied.

File: packages/core/src/facade.ts [hunk 8]
```diff
@@ -317,20 +317,249 @@ function finalizeParsedForContext(
...
-        return stripDeferredPromptValuesForContexts(descriptor.value, seen);
+        const val = stripDeferredPromptValuesForContexts(
+          descriptor.value,
+          seen,
+        );
+        if (typeof val === "function") {
+          // Own function-valued properties (class fields, constructor-
+          // assigned methods) need the sanitized-target wrapper just like
+          // prototype methods.  Cached and invalidated when the underlying
+          // function changes.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:514]: Fixed in d19842f. The proxy `get` trap now returns the exact value for non-configurable non-writable own function properties, preserving the proxy invariant.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 9]
```diff
@@ -317,20 +317,258 @@ function finalizeParsedForContext(
...
+        if (stripped !== desc.value) {
+          try {
+            Object.defineProperty(target, key, { ...desc, value: stripped });
+            saved.set(key, desc);
+            sanitizedValues.set(key, stripped);
+          } catch {
+            // Property is non-configurable or object is frozen; cannot
+            // safely call the method with unsanitized state.
+            for (const [k, d] of saved) {
+              Object.defineProperty(target, k, d);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Fixed in 3a30c1b. The rollback loop is now wrapped in try/catch per property.

File: packages/config/src/index.ts [hunk 10]
```diff
@@ -183,20 +183,234 @@ function containsDeferredPromptValues(
...
+      if (desc != null && "value" in desc) {
+        const stripped = strip(desc.value, seen);
+        if (stripped !== desc.value) {
+          try {
+            Object.defineProperty(target, key, { ...desc, value: stripped });
+            saved.set(key, desc);
+            sanitizedValues.set(key, stripped);
+          } catch {
+            for (const [k, d] of saved) {
+              Object.defineProperty(target, k, d);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts]: Fixed in 3a30c1b. Same fix applied to the config path.

File: packages/core/src/facade.ts [hunk 11]
```diff
@@ -317,20 +317,258 @@ function finalizeParsedForContext(
...
+          // assigned methods) need the sanitized-target wrapper just like
+          // prototype methods.  Cached and invalidated when the underlying
+          // function changes.
+          const cached = methodCache.get(key);
+          if (cached != null && cached.fn === val) return cached.wrapper;
+          const wrapper = function (this: unknown, ...args: unknown[]) {
+            if (this !== proxy) {
+              return stripDeferredPromptValuesForContexts(
+                val.apply(this, args),
+                seen,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:521]: Fixed in 14f6af3. Class constructors detected via `Function.prototype.toString` are now returned unwrapped, preserving `new.target` and prototype chain semantics.

File: packages/core/src/facade.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Correction: the actual commit hash is 73332e1, not 3a30c1b as stated in my previous reply.

File: packages/config/src/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts]: Correction: the actual commit hash is 73332e1.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 12]
```diff
@@ -317,20 +317,267 @@ function finalizeParsedForContext(
...
+      ) {
+        const d = Object.getOwnPropertyDescriptor(proto, key);
+        if (d != null) {
+          isAccessor = "get" in d;
+          break;
+        }
+      }
+      const result = Reflect.get(target, key, receiver);
+      if (typeof result === "function") {
+        if (!isAccessor) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:562]: Fixed in be9889f. Non-function values from accessors/prototype are now passed through `stripDeferredPromptValuesForContexts()` before returning.

File: packages/config/src/index.ts [hunk 13]
```diff
@@ -183,20 +183,243 @@ function containsDeferredPromptValues(
...
+      ) {
+        const d = Object.getOwnPropertyDescriptor(proto, key);
+        if (d != null) {
+          isAccessor = "get" in d;
+          break;
+        }
+      }
+      const result = Reflect.get(target, key, receiver);
+      if (typeof result === "function") {
+        if (!isAccessor) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts:402]: Fixed in be9889f. Same fix applied to the config path.

File: packages/core/src/facade.ts [hunk 14]
```diff
@@ -317,20 +317,267 @@ function finalizeParsedForContext(
...
+          };
+          methodCache.set(key, { fn: result, wrapper });
+          return wrapper;
+        }
+        // Accessor-returned function: wrap without caching.
+        return function (this: unknown, ...args: unknown[]) {
+          if (this !== proxy) {
+            return stripDeferredPromptValuesForContexts(
+              result.apply(this, args),
+              seen,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:589]: Fixed in be9889f. Class constructors from prototype/accessor are now detected via `Function.prototype.toString` and returned unwrapped, matching the own-property path.

**Comment by dahlia** [comment]
@codex review

File: CHANGES.md [hunk 15]
```diff
@@ -363,7 +370,7 @@ To be released.
 [#548]: https://github.com/dahlia/optique/pull/548
 [#553]: https://github.com/dahlia/optique/pull/553
 [#554]: https://github.com/dahlia/optique/pull/554
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md:378]: Fixed in 7ae861f. Restored the missing [\#555] link definition.

File: packages/core/src/facade.ts [hunk 16]
```diff
@@ -317,26 +317,283 @@ function finalizeParsedForContext(
...
-  // FIXME: Proxying non-plain parsed values changes method receiver semantics,
-  // and method-only wrappers still need a safe way to expose scrubbed values.
-  // See: https://github.com/dahlia/optique/issues/407
-  const proxy = new Proxy(value, {
+  // NOTE: Methods are invoked with the proxy as receiver so that property
+  // reads inside the method body go through the sanitized get trap.  If the
+  // method accesses private fields, that call throws a TypeError because
+  // private fields are bound to the original instance, not the proxy; in
+  // that case we temporarily sanitize the target's own properties, retry
+  // with the target as receiver, and restore the original values afterward.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Fixed in b3e11f5. Updated the NOTE to describe the current callMethodOnSanitizedTarget approach.

File: packages/config/src/index.ts [hunk 17]
```diff
@@ -183,26 +183,255 @@ function containsDeferredPromptValues(
...
-  // FIXME: This proxy changes method receiver semantics and still cannot scrub
-  // deferred prompt values hidden behind private fields or wrapper methods.
-  // See: https://github.com/dahlia/optique/issues/407
-  const proxy = new Proxy(value, {
+  // NOTE: Methods are invoked with the proxy as receiver so that property
+  // reads inside the method body go through the sanitized get trap.  If the
+  // method accesses private fields, that call throws a TypeError because
+  // private fields are bound to the original instance, not the proxy; in
+  // that case we temporarily sanitize the target's own properties, retry
+  // with the target as receiver, and restore the original values afterward.
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts]: Fixed in b3e11f5. Same update applied to the config path.

File: packages/core/src/facade.ts [hunk 18]
```diff
@@ -317,26 +317,283 @@ function finalizeParsedForContext(
...
+        try {
+          // Only restore if the method did not mutate or delete the
+          // property.  Compare against the exact sanitized value that was
+          // set, not a re-computed strip() which would create a new clone.
+          const current = Object.getOwnPropertyDescriptor(target, key);
+          if (current == null) continue; // method deleted the property
+          if (
+            "value" in current &&
+            current.value !== active!.sanitizedValues.get(key)
+          ) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:400]: This is a known limitation documented in the PR description. The sanitized value is typically `undefined`, and distinguishing an explicit `this.field = undefined` from "untouched" would require additional bookkeeping that adds complexity for an extremely narrow edge case.

File: packages/config/src/index.ts [hunk 19]
```diff
@@ -183,26 +183,255 @@ function containsDeferredPromptValues(
...
+    active!.count--;
+    if (active!.count === 0) {
+      activeSanitizations.delete(target);
+      for (const [key, desc] of active!.saved) {
+        try {
+          const current = Object.getOwnPropertyDescriptor(target, key);
+          if (current == null) continue;
+          if (
+            "value" in current &&
+            current.value !== active!.sanitizedValues.get(key)
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts:259]: Same as above; this is documented as a known limitation in the PR description.

File: packages/core/src/facade.ts [hunk 20]
```diff
@@ -317,26 +317,283 @@ function finalizeParsedForContext(
...
+    // increment the reference count so we defer restoration.
+    active.count++;
+  } else {
+    const saved = new Map<PropertyKey, PropertyDescriptor>();
+    const sanitizedValues = new Map<PropertyKey, unknown>();
+    for (const key of Reflect.ownKeys(target)) {
+      const desc = Object.getOwnPropertyDescriptor(target, key);
+      if (desc != null && "value" in desc) {
+        const stripped = strip(desc.value, seen);
+        if (stripped !== desc.value) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:362]: Fixed in 9f2a3b8. The `strip()` call is now wrapped in try/catch with rollback on failure.

File: packages/core/src/facade.ts [hunk 21]
```diff
@@ -317,26 +317,283 @@ function finalizeParsedForContext(
...
+          return result;
+        }
+        if (!isAccessor) {
+          const cached = methodCache.get(key);
+          if (cached != null && cached.fn === result) return cached.wrapper;
+          const wrapper = function (this: unknown, ...args: unknown[]) {
+            if (this !== proxy) {
+              return stripDeferredPromptValuesForContexts(
+                result.apply(this, args),
+                seen,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:569]: The class constructor check at the top of the `typeof result === 'function'` block already covers both cached and accessor paths. Traditional constructable functions (non-class) are indistinguishable from regular functions by `toString()` and remain wrapped.

File: packages/core/src/facade.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Correction: the actual commit hash is 8abc07c.

File: packages/config/src/index.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts]: Correction: the actual commit hash is 8abc07c.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.test.ts [hunk 22]
```diff
@@ -8260,4 +8260,674 @@ describe("branch coverage: facade.ts edge cases", () => {
...
+      () =>
+        runWith(parser, "test", [dynamicContext], {
+          args: [],
+        }),
+      { name: "TypeError", message: "private key is invalid." },
+    );
+
+    // The method runs through both the facade and config proxies,
+    // but must not be retried by either.
+    assert.ok(callCount <= 2);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts]: Fixed in 7f43abf. Tightened to `assert.equal(callCount, 1)`.

File: packages/core/src/facade.test.ts [hunk 23]
```diff
@@ -8260,4 +8260,674 @@ describe("branch coverage: facade.ts edge cases", () => {
...
+
+    await assert.rejects(
+      () =>
+        runWith(parser, "test", [dynamicContext], {
+          args: [],
+        }),
+      { name: "TypeError", message: "Cannot access private endpoint." },
+    );
+
+    assert.ok(callCount <= 2);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts]: Fixed in 7f43abf. Same tightening applied.

File: packages/core/src/facade.ts [hunk 24]
```diff
@@ -317,26 +317,295 @@ function finalizeParsedForContext(
...
+      // Accessor properties and prototype methods: resolved via Reflect.get.
+      // Only prototype data methods are cached; accessor-returned functions
+      // are re-created on each access since the getter may return different
+      // functions based on backing state.  Walk the prototype chain to
+      // detect inherited getters, not just own accessors.
+      let isAccessor = false;
+      for (
+        let proto: object | null = target;
+        proto != null;
+        proto = Object.getPrototypeOf(proto)
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:547]: Acknowledged. This is a valid optimization but not a correctness issue. The prototype chain walk is only reached for non-own properties, and method calls on sanitized views are not expected to be on hot paths (they run during config loading, not in tight loops). Could be addressed in a follow-up if profiling shows it matters.

File: packages/config/src/index.ts [hunk 25]
```diff
@@ -183,26 +183,266 @@ function containsDeferredPromptValues(
...
+        if (d != null) {
+          isAccessor = "get" in d;
+          break;
+        }
+      }
+      const result = Reflect.get(target, key, receiver);
+      if (typeof result === "function") {
+        if (/^class[\s{]/.test(Function.prototype.toString.call(result))) {
+          return result;
+        }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts:401]: Same as above; valid optimization, not a correctness issue in the current usage context.

File: packages/core/src/facade.ts [hunk 26]
```diff
@@ -317,26 +317,295 @@ function finalizeParsedForContext(
...
+          // Only restore if the method did not mutate or delete the
+          // property.  Compare against the exact sanitized value that was
+          // set, not a re-computed strip() which would create a new clone.
+          const current = Object.getOwnPropertyDescriptor(target, key);
+          if (current == null) continue; // method deleted the property
+          if (
+            "value" in current &&
+            current.value !== active!.sanitizedValues.get(key)
+          ) {
+            continue; // method wrote a new value
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:401]: This is an extremely narrow edge case: a method called through the sanitized proxy would need to intentionally convert a data property to an accessor during execution. This is not something config loaders are expected to do. Documenting as a known limitation rather than adding complexity for this scenario.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 27]
```diff
@@ -317,26 +317,295 @@ function finalizeParsedForContext(
...
+        );
+        if (typeof val === "function") {
+          // For non-configurable non-writable properties, the proxy invariant
+          // requires returning the exact value.  Class constructors are also
+          // returned unwrapped since the wrapper would break new.target and
+          // prototype chain semantics when invoked with `new`.
+          if (
+            (!descriptor.configurable && !descriptor.writable) ||
+            /^class[\s{]/.test(Function.prototype.toString.call(val))
+          ) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:508]: Fixed in 13529dd. The get trap now returns the exact value for non-configurable non-writable own data properties.

File: packages/config/src/index.ts [hunk 28]
```diff
@@ -183,26 +183,266 @@ function containsDeferredPromptValues(
...
+              target,
+              args,
+              stripDeferredPromptValues,
+              seen,
+            );
+          };
+          methodCache.set(key, { fn: val, wrapper });
+          return wrapper;
+        }
+        return val;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts:383]: Fixed in 13529dd. Same fix applied to the config path.

File: packages/core/src/facade.ts [hunk 29]
```diff
@@ -317,26 +317,295 @@ function finalizeParsedForContext(
...
+        if (d != null) {
+          isAccessor = "get" in d;
+          break;
+        }
+      }
+      const result = Reflect.get(target, key, receiver);
+      if (typeof result === "function") {
+        // Class constructors are returned unwrapped since the wrapper
+        // would break new.target and prototype chain semantics.
+        if (/^class[\s{]/.test(Function.prototype.toString.call(result))) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:559]: Traditional constructable functions (`function Foo(){}`) are indistinguishable from regular functions by `toString()`. This is documented as a known limitation. Adding heuristics like checking `.prototype` would incorrectly skip wrapping for all non-arrow functions.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 30]
```diff
@@ -317,26 +317,300 @@ function finalizeParsedForContext(
...
+
+interface ActiveSanitization {
+  readonly saved: Map<PropertyKey, PropertyDescriptor>;
+  readonly sanitizedValues: Map<PropertyKey, unknown>;
+  count: number;
+}
+
+const activeSanitizations = new WeakMap<object, ActiveSanitization>();
+
+function callWithSanitizedOwnProperties(
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:330]: Agreed that the duplication is not ideal. However, extracting shared helpers would require adding a new internal subpath export to @optique/core and is better suited as a follow-up refactoring. The two implementations intentionally use different strip functions (`stripDeferredPromptValuesForContexts` vs `stripDeferredPromptValues`) which makes a simple extraction non-trivial.

File: packages/config/src/index.ts [hunk 31]
```diff
@@ -183,26 +183,271 @@ function containsDeferredPromptValues(
...
+const SANITIZE_FAILED: unique symbol = Symbol("sanitizeFailed");
+
+interface ActiveSanitization {
+  readonly saved: Map<PropertyKey, PropertyDescriptor>;
+  readonly sanitizedValues: Map<PropertyKey, unknown>;
+  count: number;
+}
+
+const activeSanitizations = new WeakMap<object, ActiveSanitization>();
+
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/config/src/index.ts:195]: Same as above; follow-up refactoring.

File: packages/core/src/facade.ts [hunk 32]
```diff
@@ -317,26 +317,300 @@ function finalizeParsedForContext(
...
+          // Only restore if the method did not mutate or delete the
+          // property.  Compare against the exact sanitized value that was
+          // set, not a re-computed strip() which would create a new clone.
+          const current = Object.getOwnPropertyDescriptor(target, key);
+          if (current == null) continue; // method deleted the property
+          if (
+            "value" in current &&
+            current.value !== active!.sanitizedValues.get(key)
+          ) {
+            continue; // method wrote a new value
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:401]: This is the same edge case noted in the PR description under known limitations. Methods that convert data properties to accessors during execution are not expected in config loading contexts. The additional complexity to track descriptor shape changes is not warranted for this narrow case.
